experiments/billable/src/worker.tsx

```diff
@@ -144,86 +89,110 @@ export default {
-        } else {
-          return new Response('No access', { status: 401 });
-        }
-      }
-
-
-      if (pathname === "/") {
-        return new Response(`
-          <html>
-            <head>
-              <meta http-equiv="refresh" content="0;url=/invoices">
-            </head>
-            <body>
-              Redirecting to invoices...
-            </body>
-          </html>`, {
-          status: 200,
-          headers: {
-            "Content-Type": "text/html"
-          },
-        });
-      }
-
-      if (Page) {
-        return renderPage(Page, { ctx });
-      }
-
-      if (pathname.startsWith("/invoice/")) {
-        const id = pathname.slice("/invoice/".length);
-        if (pathname.endsWith("/upload")) {
-          if (
-            request.method === "POST" &&
-            request.headers.get("content-type")?.includes("multipart/form-data")
-          ) {
-            const formData = await request.formData();
-            const userId = formData.get("userId") as string;
-            const invoiceId = formData.get("invoiceId") as string;
-            const file = formData.get("file") as File;
-
-            // Stream the file directly to R2
-            const r2ObjectKey = `/logos/${userId}/${invoiceId}-${Date.now()}-${file.name}`;
-            await env.R2.put(r2ObjectKey, file.stream(), {
-              httpMetadata: {
-                contentType: file.type,
+      const r = defineRoutes(
+        [
+          index(HomePage),
+
+          route("/login", LoginPage),
+          route("/auth", async (req) => {
+            // when it's async then react-is thinks it's a react component.
+            const url = new URL(req.url);
+            const token = url.searchParams.get("token");
+            const email = url.searchParams.get("email");
+
+            if (!token || !email) {
+              return new Response("Invalid token or email", { status: 400 });
+            }
+            const user = await db.user.findFirst({
+              where: {
+                email,
+                authToken: token,
+                authTokenExpiresAt: {
+                  gt: new Date(),
+                },
               },
             });
 
-            await db.invoice.update({
-              where: { id: invoiceId },
+            if (!user) {
+              return new Response("Invalid or expired token", { status: 400 });
+            }
+
+            // Clear the auth token
+            await db.user.update({
+              where: { id: user.id },
               data: {
-                supplierLogo: r2ObjectKey,
+                authToken: null,
+                authTokenExpiresAt: null,
               },
             });
 
-            return new Response(JSON.stringify({ key: r2ObjectKey }), {
-              status: 200,
+            console.log("performing login");
+
+            return performLogin(request, env, user.id);
+          }),
+
+          route("/invoices", InvoiceListPage),
```

Comment:
@peterp so the handlers for routes are just components? That's awesome!!!

Reply:
Yeah!